Sources/SwiftParserDiagnostics/ParserDiagnosticMessages.swift

```diff
@@ -305,6 +296,14 @@ public struct CannotParseVersionTuple: ParserError {
   }
 }
 
+public struct DeinitializerSignatureError: ParserError {
+  public let nodes: [SyntaxProtocol]
+
+  public var message: String {
+    return "deinitializers cannot have \(nodes.map { $0.nodeTypeNameForDiagnostics(allowBlockNames: false) ?? "syntax" }.joined(separator: " or "))"
```

Comment:
Could you just call `nodesDescription(nodes)` instead here? That should call into `childNameForDiagnostics`. And if you set the `nameForDiagnostics` of `DeinitializerSyntax.name`, I would imagine that this should give you a decent error message.

Reply:
The thing is `DeinitializerSyntax` don't have a name 😅
Updated all the test cases so you can see the result

Reply:
Ah, sorry, I forgot that. 

What about making `DeinitializerSignatureError` have three children `name: TokenSyntax?`, `params: FunctionParameterClauseSyntax?` and `returnClause: ReturnClauseSyntax?`. And then `DeinitializerSignatureError` would stitch together the works `name`, `parameters` and `return clause`.